user_tools/src/spark_rapids_tools/storagelib/tools/fs_utils.py

```diff
@@ -0,0 +1,186 @@
+    existing file.
+    ```py
+    try:
+        file_algo = FileHashAlgorithm(algorithm=HashAlgorithm.SHA256, value='...')
+        file_algo.verify_file(CspPath('file://path/to/file'))
+    except ValidationError as e:
+        print(e)
+    ```
+    """
+    algorithm: HashAlgorithm
+    value: str
+
+    def verify_file(self, file_path: CspPath) -> bool:
+        cb = self.algorithm.get_hash_func()
+        with file_path.open_input_stream() as stream:
+            res = cb(stream.readall())
+            if res.hexdigest() != self.value:
+                raise_invalid_file(file_path, f'incorrect file hash [{HashAlgorithm.tostring(self.algorithm)}]')
+        return True
+
+
+@dataclasses.dataclass
+class FileVerificationResult:
+    """
+    A class that represents the result of a file verification.
+    :param res_path: The path to the resource subject of verification.
+    :param opts: Checking options to be passed to the FileChecker
+    :param raise_on_error: Flag to raise an exception if the file is invalid.
+    """
+    res_path: CspPath
+    opts: dict
+    raise_on_error: bool = False
+    validation_error: Optional[ValidationError] = dataclasses.field(default=None, init=False)
+
+    def __post_init__(self):
+        try:
+            CspFileChecker(**{'file_path': str(self.res_path), **self.opts})
+        except ValidationError as err:
+            self.validation_error = err
+            if self.raise_on_error:
+                raise err
+
+    @cached_property
+    def successful(self) -> bool:
+        return self.validation_error is None
+
+
+@dataclass
+class CspFileChecker:
+    """
+    A class that represents a file checker. It is used as a pydantic model to validate the file.
+    :param file_path: The path to the file. It accepts any valid CspPath or AnyHttpUrl in case we want to
+           verify an url path extension even before we download it.
+    :param must_exist: When True, the file must exist.
+    :param is_file: When True, the file must be a file. Otherwise, it is a directory.
+    :param size: The expected size of the file. When 0, it is not checked.
+    :param extensions: A list of expected extensions. When None, it is not checked.
+    :param file_hash: A hash algorithm and its value to verify the file.
+    ```py
+    try:
+        TypeAdapter(CspFileChecker).validate_python({
+            'file_path': 'file:///var/tmp/spark_cache_folder_test/rapids-4-spark-tools_2.12-24.08.2.jar',
+            'must_exist': True,
+            'size': .....,
+            'extensions': ['jar']})
+    except ValidationError as e:
+        print(e)
+    ```
+    """
+    file_path: Union[CspPathString, FilePath, AnyHttpUrl]
+    must_exist: Optional[bool] = False
+    is_file: Optional[bool] = True
+    size: Optional[int] = 0
+    extensions: Optional[List[str]] = None
+    file_hash: Optional[FileHashAlgorithm] = None
+    # TODO add verification for the modified time.
```

Comment:
is there a followup for this?

Reply:
No, I did not file a separate followup for this, but I will add it to subtasks in #1286.
We used to have modified-time verification because we would refresh the cached files every 2 weeks. That was mainly used in the "Price catalogs".